src/dsl/compute_graph.jl

```diff
@@ -146,17 +208,282 @@ ncon c 1,4 a 1,2 b 2,1,4
+        empty!(node.data.left_idxs)
+        append!(node.data.left_idxs, new_left_idxs)
+    end
+    if isdefined(node, :right)
+        r_index_map = Dict(x => i for (i, x) in enumerate(node.data.right_idxs))
+        m = x -> [r_index_map[y] for y in x]
+        right_remaining_idxs = filter(x -> x in node.data.right_idxs, remaining_idxs)
+        # right_remaining_idx_groups = find_overlaps(node.data.output_idxs, right_remaining_idxs)
+        new_index_order = length(common_idxs) == 0 ? Vector{Vector{Int}}() : [m(common_idxs)]
+        append!(new_index_order, [map(x -> [x], m(right_remaining_idxs))..., map(x -> [x], m(batch_idxs))...])
+        # append!(new_index_order, [m.(right_remaining_idx_groups)..., map(x -> [x], m(batch_idxs))...])
+        permute_and_merge!(node.right, new_index_order)
+        new_right_idxs = length(common_idxs) > 0 ? [common_idxs[1]] : Int[]
+        append!(new_right_idxs, [right_remaining_idxs..., batch_idxs...])
+        empty!(node.data.right_idxs)
+        append!(node.data.right_idxs, new_right_idxs)
+    end
+    nothing
+end
+
+batched_indices(cmd) = intersect(cmd.output_idxs, cmd.left_idxs, cmd.right_idxs)
+
+function _permute_and_merge!(cmd::ContractCommand, new_index_order)
+    if new_index_order !== nothing
+        # get mapping between indices as seen by next command, ordered 1,2..n where n is rank
+        # and grouped according to reshape_groups
+        current_idxs = Dict{Int, Vector{Int}}()
+        pos = 1
+        for (i, group_size) in enumerate(cmd.reshape_groups)
+            current_idxs[i] = cmd.output_idxs[pos:pos+group_size-1]
+            pos += group_size
+        end
+        flat_order = vcat(new_index_order...)
+        # @show cmd.output_name, new_index_order
+        # @show cmd.output_idxs, vcat(map(x -> current_idxs[x], flat_order)...)
+        cmd.output_idxs[:] = vcat(map(x -> current_idxs[x], flat_order)...)
+        # update reshape groups to reflect any new merges
+        new_reshape_groups = []
+        pos = 1
+        for l in length.(new_index_order)
+            push!(new_reshape_groups, sum(cmd.reshape_groups[pos:pos+l-1]))
+            pos += l
+        end
+        empty!(cmd.reshape_groups)
+        append!(cmd.reshape_groups, new_reshape_groups)
+    end
+end
+
+"""
+    join_remaining!(node::ComputeNode{ContractCommand}, new_index_order=nothing)
+
+Identify groups of indices that are grouped in the output indices and also appear in the same
+order in the left and right index groups. Merge these in left and right index groups and pass
+this grouping to children
+
+Example:
+
+"""
+function join_remaining!(node::ComputeNode{ContractCommand}, new_index_order=nothing)
+    _join_remaining!(node.data, new_index_order)
+
+    # find output indices that are merged in the reshape after the contraction
+    pos = 1
+    output_groups = Vector{Vector{Int}}()
+    for j in node.data.reshape_groups
+        push!(output_groups, node.data.output_idxs[pos:pos+j-1])
```

Comment:
A very similar loop appears in the `write` and `_permute_and_merge` functions so I'm wondering if a `get_groups` function would be worth adding. The other loops seem slightly different also though so I'm not sure it would be of any benefit.